rest-service/manager_rest/resource_manager.py

```diff
@@ -201,38 +201,50 @@ def update_execution_status(self, execution_id, status, error):
-                    same_deployment_executions.append(execution)
-                else:
-                    other_executions.append(execution)
-        else:
-            other_executions = queued_executions
-
-        # same-deployment executions run first, no matter if there's system
-        # executions or not
-        for e in same_deployment_executions:
-            self.execute_queued_workflow(e)
-
-        for e in other_executions:
-            self.execute_queued_workflow(e)
-            if e.is_system_workflow:  # To avoid starvation of system workflows
-                break
-
-    def _get_queued_executions(self):
-        filters = {'status': ExecutionState.QUEUED_STATE}
-
-        queued_executions = self.list_executions(
-            is_include_system_workflows=True,
-            filters=filters,
-            all_tenants=True,
-            sort={'created_at': 'asc'}).items
+        for execution in self._get_queued_executions(finished_execution):
+            if self._should_use_system_workflow_executor(execution):
+                self._execute_system_workflow(execution, queue=True)
+            else:
+                self.execute_workflow(execution, queue=True)
+
+    def _get_queued_executions(self, finished_execution):
+        system_executions = self.sm.list(models.Execution, filters={
+            'status': ExecutionState.QUEUED_STATE,
+            'is_system_workflow': True,
+        }, sort={'created_at': 'asc'}, get_all_results=True).items
+        if system_executions:
+            yield system_executions[0]
+            return
 
-        return queued_executions
+        deployment_id = finished_execution.deployment.id
+        if finished_execution.deployment:
+            same_dep_executions = self.sm.list(models.Execution, filters={
+                'status': ExecutionState.QUEUED_STATE,
+                'deployment_id': deployment_id,
+            }, sort={'created_at': 'asc'}, get_all_results=True).items
+            if same_dep_executions:
+                yield same_dep_executions[0]
+
+        other_queued = self.sm.list(models.Execution, filters={
+            'status': ExecutionState.QUEUED_STATE,
+            'deployment_id': lambda col: col != deployment_id,
+        }, sort={'created_at': 'asc'}, get_all_results=True)
+
+        seen_deployments = set()
+        group_counts = {}
+        for execution in other_queued:
+            for group in execution.execution_groups:
+                if group.id not in group_counts:
+                    group_counts[group.id] = group.concurrency -\
+                        len(group.currently_running_executions())
+        for execution in other_queued:
+            if execution.deployment in seen_deployments:
+                continue
+            if all(group_counts[g.id] > 0 for g in execution.execution_groups):
+                for group in execution.execution_groups:
+                    group_counts[group.id] -= 1
```

Comment:
why?

Reply:
group_counts is the "how many executions can this group still run"

here, we're going to be running an execution, so we're reserving a "slot" in each of its groups

Reply:
just to clarify:
* `system_executions` do not need any slots (because are not part of any execution group?);
* `same_dep_executions` because the new execution is part of the same execution group as the finished one.
right?

Reply:
oooh good point, same-dep needs to be under a group check as well